### PR TITLE
Test re-syncing state from other servers listed in `/send_join` reply

### PR DESCRIPTION
**Pull request by squahtx**
Tests https://github.com/matrix-org/synapse/pull/12812

### CODE REVIEW

File: internal/federation/handle.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ func MakeJoinRequestsHandler(s *Server, w http.ResponseWriter, req *http.Request
 		Type:       "m.room.member",
 		StateKey:   &userID,
 		PrevEvents: []string{room.Timeline[len(room.Timeline)-1].EventID()},
+		Depth:      room.Timeline[len(room.Timeline)-1].Depth() + 1,
```
Comment (squahtx) [inline internal/federation/handle.go:48]: This is a drive-by fix and doesn't affect the outcome of the test.

File: internal/federation/server_room.go [hunk 2]
```diff
@@ -97,24 +97,38 @@ func (r *ServerRoom) AuthChain() (chain []*gomatrixserverlib.Event) {
...
 	}
-	// find them in the timeline
-	for _, tev := range r.Timeline {
-		if chainMap[tev.EventID()] {
-			chain = append(chain, tev)
-		}
-	}
+
 	return
 }
```
Comment (squahtx) [inline internal/federation/server_room.go:133]: Without this change, the complement homeserver excludes hs1's membership event from the auth chain sent to hs2. hs2 is unhappy and tries to make `/event_auth` requests, which the complement homeserver is not set up to handle.

File: internal/federation/handle.go (hunk 1 above)
Comment (kegsay) [reply internal/federation/handle.go:48]: Please don't mix this with tests where possible. The set of reviewers for $feature are different from the set of reviewers for handling Complement internals.

File: internal/federation/server.go [hunk 3]
```diff
@@ -119,6 +119,11 @@ func NewServer(t *testing.T, deployment *docker.Deployment, opts ...func(*Server
 	return srv
 }
 
+// Room returns the given room
+func (s *Server) Room(roomID string) *ServerRoom {
```
Comment (kegsay) [inline internal/federation/server.go:123]: This is not required. You call `MustJoinRoom` in the test which returns the `*ServerRoom` already.

File: tests/federation_room_join_partial_state_test.go [hunk 4]
```diff
@@ -182,6 +182,104 @@ func TestPartialStateJoin(t *testing.T) {
...
+				},
+				// hs1 may send us presence when alice syncs
+				func(e gomatrixserverlib.EDU) {},
+			),
+		)
+		cancelListener := server.Listen()
+		defer cancelListener()
+
+		// join complement to the public room
+		server.MustJoinRoom(t, deployment, "hs1", roomID, server.UserID("bob"))
```
Comment (kegsay) [inline tests/federation_room_join_partial_state_test.go:217]: This returns a `*ServerRoom` so you don't need to then retrieve it with the new `server.Room(roomID)` function on `:245`.

File: tests/federation_room_join_partial_state_test.go [hunk 5]
```diff
@@ -182,6 +182,104 @@ func TestPartialStateJoin(t *testing.T) {
...
+
+		// the client-side requests should still be waiting
+		select {
+		case <-syncResponseChan:
+			t.Fatalf("hs2 sync completed before state resync complete")
+		default:
+		}
+
+		// reply to hs2 with a bogus /state_ids response
+		sendResponseWaiter.Finish()
```
Comment (kegsay) [inline tests/federation_room_join_partial_state_test.go:268]: This could have a clearer name, e.g `sendStateIDsResponse`, as otherwise it's unclear which response you mean and need to add comments to explain it.

**Review by kegsay — APPROVED** [review]
Overall this is an excellent PR! Thanks!

File: tests/federation_room_join_partial_state_test.go (hunk 5 above)
Comment (squahtx) [reply tests/federation_room_join_partial_state_test.go]: I'll rename them to follow the naming in `partialStateJoinResult`:
```go
type partialStateJoinResult struct {
	cancelListener                   func()
	Server                           *federation.Server
	ServerRoom                       *federation.ServerRoom
	fedStateIdsRequestReceivedWaiter *Waiter
	fedStateIdsSendResponseWaiter    *Waiter
}
```

**Comment by squahtx** [comment]
@kegsay Thanks for reviewing. Could you re-review the latest changes?

File: tests/federation_room_join_partial_state_test.go [hunk 6]
```diff
@@ -182,6 +182,103 @@ func TestPartialStateJoin(t *testing.T) {
...
+		server := federation.NewServer(t, deployment,
+			federation.HandleKeyRequests(),
+			federation.HandlePartialStateMakeSendJoinRequests(),
+			federation.HandleEventRequests(),
+			federation.HandleTransactionRequests(
+				func(e *gomatrixserverlib.Event) {
+					t.Fatalf("Received unexpected PDU: %s", string(e.JSON()))
+				},
+				// hs1 may send us presence when alice syncs
+				func(e gomatrixserverlib.EDU) {},
```
Comment (richvdh) [inline tests/federation_room_join_partial_state_test.go:210]: I think you can just write `nil` if there's nothing to do for EDUs.
```suggestion
				nil,
```

File: tests/federation_room_join_partial_state_test.go [hunk 7]
```diff
@@ -182,6 +182,103 @@ func TestPartialStateJoin(t *testing.T) {
...
+			),
+		)
+		cancelListener := server.Listen()
+		defer cancelListener()
+
+		// join complement to the public room
+		room := server.MustJoinRoom(t, deployment, "hs1", roomID, server.UserID("bob"))
```
Comment (richvdh) [inline tests/federation_room_join_partial_state_test.go:217]: worth noting there is also a `bob@hs1` created by the blueprint. It might be better to pick a fourth username.

**Review by richvdh — APPROVED** [review]
lgtm other than a few tweaks

File: tests/federation_room_join_partial_state_test.go (hunk 7 above)
Comment (squahtx) [reply tests/federation_room_join_partial_state_test.go]: I went for `david` instead.

**Review by richvdh — APPROVED** [review]
lgtm, and I need to build stuff on top of it, so I'm going to go ahead and merge it.
